---
title: "RAMP Metagenomics Analysis"
output: html_document
date: '2022-08-23'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Maaslin2)
library(vegan)
library(data.table)

select = dplyr::select
rename = dplyr::rename
```

## Load metadata

```{r load_metadata}
bifido_responders = read_csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/16S/16S_bifido_responder_status.csv') %>%
  mutate(participant_id = factor(participant_id))

final_participant_list = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/metadata/RAMP_final_participant_list.csv')

ramp_metadata_df = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/metadata/RAMP_participant_metadata.csv') %>%
  mutate(participant_id = factor(participant_id),
         final_grp = factor(final_grp, levels=c('Placebo', 'Low Dose 2FL', 'High Dose 2FL'))) %>%
  left_join(bifido_responders) %>%
  filter(participant_id %in% final_participant_list$participant_id)

final_participant_timepoint = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/metadata/RAMP_final_participant_timepoint.csv') 

ramp_metadata_all_timepoints = final_participant_timepoint %>% 
  rowwise() %>%
  mutate(participant_id = strsplit(participant_timepoint, '_')[[1]][1],
         timepoint = strsplit(participant_timepoint, '_')[[1]][2]) %>%
  left_join(ramp_metadata_df) %>%
  mutate(participant_id = factor(participant_id),
         final_grp = factor(final_grp, levels=c('Placebo', 'Low Dose 2FL', 'High Dose 2FL')))

stool_manifest = read_csv('/home/mmcarter/user_data/Projects/RAMP/METAGENOMICS/220113_stool_manifest.csv') %>% 
  rename(participant_id = 1,
         visit_date = 2,
         scfa_index = 6) %>%
  select(participant_id, visit_date, stool_count, scfa_index, `MGX Plate`) %>%
  rowwise() %>%
  mutate(participant_id = factor(participant_id)) %>%
  left_join(ramp_metadata_df, by='participant_id') %>%
  mutate(joined_sample_id = paste0(participant_id, '__', str_replace_all(visit_date, '/', '_')))


```

## Load QC data

```{r load_qc}
table_parse = function(path) {
  sample_id = strsplit(basename(path), '__')[[1]][1]
  return(fread(path, sep='\t') %>% mutate(sample_id = sample_id))
}

qc_summary_file_path = '/home/mmcarter/user_data/scripts/210813_preprocessing_ramp_mgx/qc_summaries/'
files <- list.files(path = qc_summary_file_path, pattern = ".txt")
qc_summary_df <- lapply(paste0(qc_summary_file_path, files), table_parse) %>% 
  rbindlist() %>%
  arrange(desc(raw_bases)) %>%
  filter(SampleName != '4010__No_date')

mean(qc_summary_df$unique_reads)
sd(qc_summary_df$unique_reads)

mean(qc_summary_df$unique_bases)/1e9
```

## Read humann results

```{r load_humann_data}
pathabundance_unstratified_table = read_tsv(file = '/home/mmcarter/user_data/Projects/RAMP/METAGENOMICS/humann3_files/RAMP_HUMANN3_joined_pathabundance_unstratified.tsv') %>%
  rename(GeneFamily = 1) %>%
  filter(GeneFamily != 'UNMAPPED') %>%
  filter(GeneFamily != 'UNGROUPED')

kegg_unstratified_table = read_tsv(file = '/home/mmcarter/user_data/Projects/RAMP/METAGENOMICS/humann3_files/RAMP_HUMANN3_joined_keggfamilies_unstratified.tsv') %>%
  rename(GeneFamily = 1) %>%
  filter(GeneFamily != 'UNMAPPED') %>%
  filter(GeneFamily != 'UNGROUPED')

pfam_unstratified_table = read_tsv(file = '/home/mmcarter/user_data/Projects/RAMP/METAGENOMICS/humann3_files/RAMP_HUMANN3_joined_pfamfamilies_unstratified.tsv') %>%
  rename(GeneFamily = 1) %>%
  filter(GeneFamily != 'UNMAPPED') %>%
  filter(GeneFamily != 'UNGROUPED')

samples_to_remove = c('4058__10_24_2019', '4060__11_2_2019', '4093__3_9_2020', '4049__11_2_2019', '4026__8_21_2019')

pfam_unstratified_sample_by_pfam = kegg_unstratified_table %>%
  gather(key='pre_sample_id', value='RPK', -GeneFamily) %>%
  mutate(sample_id = str_replace(pre_sample_id, 'HUMANN3_', ''),
         sample_id = str_replace(sample_id, '_Abundance-RPKs', '')) %>%
  mutate(RPK = as.integer(RPK)) %>%
  select(-pre_sample_id) %>%
  mutate(RPK=log2(RPK+2^7)) %>%
  spread(key='GeneFamily', value='RPK') %>%
  filter(!sample_id %in% samples_to_remove) %>%
  column_to_rownames(var='sample_id')


```

## Make all-sample ordination

```{r ordination}
d = vegdist(pfam_unstratified_sample_by_pfam, method='euclidean')
c = cmdscale(d, k=4, eig=TRUE)

c$eig/sum(c$eig) %>% sort()


ramp_humann_pcoa_df = c$points %>%
  as.data.frame() %>%
  rename(Axis.1 = 1,
         Axis.2 = 2, 
         Axis.3 = 3, 
         Axis.4 = 4) %>%
  rownames_to_column(var='sample_id') %>%
  rowwise() %>%
  mutate(sample_id = ifelse(grepl('121-8', sample_id), str_replace(sample_id, '121-8', '4000'), sample_id)) %>%
  mutate(participant_id = strsplit(sample_id, '__')[[1]][1],
         visit_date = strsplit(sample_id, '__')[[1]][2],
         visit_date = str_replace_all(visit_date, '_', '/')) %>%
  left_join(stool_manifest, by=c('participant_id', 'visit_date')) %>%
  filter(!is.na(final_grp)) %>%
  mutate(stool_count = factor(stool_count, levels=c('1', '2', '3', '4', '5')),
         final_grp = factor(final_grp, levels=c('Placebo', 'Low Dose 2FL', 'High Dose 2FL'))) 

ramp_humann_all_sample_pcoa_plot = ggplot(ramp_humann_pcoa_df, aes(x=Axis.1, y=Axis.2, color=final_grp, shape=stool_count)) +
  geom_point(size=3, alpha=0.85) +
  scale_color_brewer(palette='Set2') +
  scale_color_manual(values=c('#939598', '#b4cce1', '#5ea1d7')) +
  ylab(paste0('PCo2')) +
  xlab(paste0('PCo1')) + 
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

ggsave(filename = '/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/mgx_humann3_all_sample_pcoa_plot.pdf',
       plot = ramp_humann_all_sample_pcoa_plot,
       width = 6, height = 4, units = c('in'))

pfam_trim_timepoint_3 = pfam_unstratified_sample_by_pfam[ramp_humann_pcoa_df %>% pull(sample_id), ] %>%
  na.omit()
pfam_dist_mat_3 <- vegdist(pfam_trim_timepoint_3, method='bray')
treatment_fact_3 = ramp_humann_pcoa_df %>% filter(sample_id %in% rownames(pfam_trim_timepoint_3)) %>% pull(final_grp)
subject_fact_3 = ramp_humann_pcoa_df %>% filter(sample_id %in% rownames(pfam_trim_timepoint_3)) %>% pull(participant_id)
adonis2(pfam_dist_mat_3 ~ treatment_fact_3 + subject_fact_3, by='terms', perm=1000)

beta_tp1 <- betadisper(pfam_dist_mat_3, treatment_fact_3)
permutest(beta_tp1)
boxplot(beta_tp1)


```

## Run Maaslin

```{r maaslin}

humann_metadata_maaslin = stool_manifest %>%
  select(-c(scfa_index)) %>%
  # mutate(mod_visit_date = paste0(
  #   str_replace(strsplit(as.character(visit_date), '-')[[1]][2], '^0', ''), '_',
  #   str_replace(strsplit(as.character(visit_date), '-')[[1]][3], '^0', ''), '_',
  #   strsplit(as.character(visit_date), '-')[[1]][1]
  #   )) %>% 
  mutate(rowname = paste0(participant_id, '__', str_replace_all(visit_date, '/', '_')),
         rowname = str_replace_all(rowname, '-', '_')) %>%
  filter(!is.na(stool_count)) %>%
  group_by(rowname) %>% 
  filter(row_number() == 1) %>%
  column_to_rownames(var='rowname') %>%
  mutate(num_time = case_when(
           stool_count == 1 ~ -2,
           stool_count == 2 ~ 0,
           stool_count == 3 ~ 3,
           stool_count == 4 ~ 6,
           stool_count == 5 ~ 10
         )) %>%
  filter(num_time %in% c(0, 6)) %>%
  mutate(grpHigh_time = ifelse(final_grp == 'High Dose 2FL', num_time, 0),
         grpLow_time = ifelse(final_grp == 'Low Dose 2FL', num_time, 0),
         grpPlacebo_time = ifelse(final_grp == 'Placebo', num_time, 0),
         grpResponder_time = ifelse(responder_status_delta == 'responder', num_time, 0),
         grpNonresponder_time = ifelse(responder_status_delta == 'nonresponder', num_time, 0),
         final_grp = factor(final_grp, levels=c('Placebo', 'Low Dose 2FL', 'High Dose 2FL')),
         grp_time = num_time*final_grp)

human_data_maaslin = pfam_unstratified_sample_by_pfam[row.names(humann_metadata_maaslin), ] %>%
  na.omit()


mas_metagenomics_resp_interact <- Maaslin2(
  input_data = human_data_maaslin,
  input_metadata = humann_metadata_maaslin,
  output = "/home/mmcarter/user_data/Projects/RAMP/METAGENOMICS/MAASLIN",
  min_abundance = 0.0,
  min_prevalence = 0.3,
  normalization = "NONE",
  transform = "NONE",
  analysis_method = "LM",
  max_significance = 0.2,
  fixed_effects = c("final_grp", "num_time"),
  random_effects = c("participant_id"),
  reference = c("final_grp,Placebo"),
  correction = "BH",
  standardize = TRUE,
  cores = 1)
```


## inStrain analysis

```{r }
genomeset_delta = read_csv('/LAB_DATA/CURRENT/CURRENT_Metagenomics_PROJECTS/2019_Hadza_Nepal_Metagenomics/191119_FINAL_PROJECT_DATA/ANALYSIS/inStrain/genome_databases/Delta_10062020/GenomeSetDelta_allGenomes_v3.csv')

table_parse = function(path) {
  sample_id = strsplit(basename(path), '__')[[1]][1]
  visit_date = strsplit(basename(path), '__')[[1]][2]
  return(fread(path, sep='\t') %>% mutate(sample_id = paste0(sample_id, '__', visit_date)))
}

instrain_file_path = '//home/mmcarter/user_data/scripts/210830_ramp_instrain_compare_prep/ramp_instrain_profile/'
files <- list.files(path = instrain_file_path, pattern = ".tsv")
instrain_summary_df <- lapply(paste0(instrain_file_path, files), table_parse) %>% 
  rbindlist(fill=TRUE) 


ramp_instrain_rel_ab = instrain_summary_df %>% 
  filter(breadth >= 0.5) %>% 
  group_by(sample_id) %>% 
  mutate(rel_ab = filtered_read_pair_count/sum(filtered_read_pair_count)) %>% 
  left_join(genomeset_delta %>% select(genome, GTDB_tax), by=c('genome'))

ramp_instrain_rel_ab_spread = ramp_instrain_rel_ab %>%
  select(sample_id, genome, rel_ab) %>%
  spread(key='genome', value='rel_ab', fill=0.) %>% 
  column_to_rownames(var='sample_id')

instrain_data_maaslin = ramp_instrain_rel_ab_spread[row.names(humann_metadata_maaslin), ] %>%
  na.omit()


mas_metagenomics_resp_interact <- Maaslin2(
  input_data = instrain_data_maaslin,
  input_metadata = humann_metadata_maaslin,
  output = "/home/mmcarter/user_data/Projects/RAMP/METAGENOMICS/MAASLIN",
  min_abundance = 0.0,
  min_prevalence = 0.1,
  normalization = "NONE",
  transform = "AST",
  analysis_method = "LM",
  max_significance = 0.2,
  fixed_effects = c("final_grp", "num_time"),
  random_effects = c("participant_id"),
  reference = c("final_grp,Placebo"),
  correction = "BH",
  standardize = TRUE,
  cores = 1)

mas_metagenomics_resp_interact$results %>% 
  arrange(pval, qval) %>% 
  left_join(genomeset_delta %>% select(Genome, GTDB_r95_taxonomy), by=c('feature'='Genome'))

```


```{r rel_ab_plotting}
stool_meta = stool_manifest %>%
  select(-c(scfa_index)) %>%
  # mutate(mod_visit_date = paste0(
  #   str_replace(strsplit(as.character(visit_date), '-')[[1]][2], '^0', ''), '_',
  #   str_replace(strsplit(as.character(visit_date), '-')[[1]][3], '^0', ''), '_',
  #   strsplit(as.character(visit_date), '-')[[1]][1]
  #   )) %>% 
  mutate(rowname = paste0(participant_id, '__', str_replace_all(visit_date, '/', '_')),
         rowname = str_replace_all(rowname, '-', '_')) %>%
  filter(!is.na(stool_count)) %>%
  group_by(rowname) %>% 
  filter(row_number() == 1)

ramp_instrain_rel_ab %>% 
  pull(sample_id) %>% unique()

tax_meta = genomeset_delta %>%
  select(genome, GTDB_tax)

ramp_rel_ab_w_meta = ramp_instrain_rel_ab %>% 
  filter(grepl('Bifido', GTDB_tax)) %>% 
  select(genome, sample_id, rel_ab) %>% 
  spread(key='genome', value='rel_ab', fill=0.) %>%
  gather(key='genome', value='rel_ab', -sample_id) %>%
  left_join(tax_meta) %>%
  left_join(stool_meta, by=c('sample_id'='rowname')) %>% 
  rowwise() %>%
  mutate(species_name = strsplit(GTDB_tax, ';')[[1]][7]) %>% 
  filter(!is.na(stool_count)) %>% 
  group_by(sample_id, species_name, final_grp, stool_count) %>% 
  summarise(sum_abund = sum(rel_ab)) %>%
  filter(species_name != 's__Bifidobacterium globosum' & species_name != 's__Bifidobacterium angulatum' & species_name != 's__Bifidobacterium catenulatum' & species_name != 's__Bifidobacterium animalis') %>%
  mutate(final_grp = factor(final_grp, levels=c('Placebo', 'Low Dose 2FL', 'High Dose 2FL')))

bifido_species_abundance_plot = ggplot(ramp_rel_ab_w_meta, aes(x=factor(stool_count), y=sum_abund*100, fill=final_grp)) +
  facet_wrap(~species_name, scales='free_y') +
  scale_fill_manual(values=c('#939598', '#b4cce1', '#5ea1d7')) +
  scale_y_continuous(name = 'Relative Abundance (%)', expand=c(0,0), limits=c(0, 14)) +
  scale_x_discrete(name='Timepoint (Weeks)', labels=c('-2', '0', '3', '6')) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(pch=21, position=position_jitterdodge()) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

ggsave('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/MGX_bifido_species_abundance.pdf', 
       plot = bifido_species_abundance_plot, 
       width = 6, height = 4)

wilcox.test(ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium adolescentis") %>% filter(stool_count == 3) %>% filter(final_grp == 'Placebo') %>% pull(sum_abund),
            ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium adolescentis")%>% filter(stool_count == 3) %>% filter(final_grp == 'High Dose 2FL')%>% pull(sum_abund))

wilcox.test(ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium adolescentis") %>% filter(stool_count == 4) %>% filter(final_grp == 'Placebo') %>% pull(sum_abund),
            ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium adolescentis")%>% filter(stool_count == 4) %>% filter(final_grp == 'High Dose 2FL')%>% pull(sum_abund))

wilcox.test(ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium pseudocatenulatum") %>% filter(stool_count == 3) %>% filter(final_grp == 'Placebo') %>% pull(sum_abund),
            ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium pseudocatenulatum")%>% filter(stool_count == 3) %>% filter(final_grp == 'High Dose 2FL')%>% pull(sum_abund))

wilcox.test(ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium pseudocatenulatum") %>% filter(stool_count == 4) %>% filter(final_grp == 'Placebo') %>% pull(sum_abund),
            ramp_rel_ab_w_meta %>% filter(species_name == "s__Bifidobacterium pseudocatenulatum")%>% filter(stool_count == 4) %>% filter(final_grp == 'High Dose 2FL')%>% pull(sum_abund))



bifido_stacked_bar_plot = ggplot(ramp_rel_ab_w_meta %>% mutate(species_name = factor(species_name, levels=c('s__Bifidobacterium bifidum', 's__Bifidobacterium infantis', 's__Bifidobacterium pseudocatenulatum', 's__Bifidobacterium adolescentis'))), aes(x=factor(stool_count), y=sum_abund*100, fill=species_name)) +
  facet_wrap(~final_grp) +
  scale_fill_manual(values=c('#546de5', '#e15f41', '#c44569', '#574b90')) +
  scale_y_continuous(name = 'Relative Abundance (%)', expand=c(0,0), limits=c(0, 45)) +
  scale_x_discrete(name='Timepoint (Weeks)', labels=c('-2', '0', '3', '6')) +
  geom_bar(stat='identity') +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

ggsave('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/MGX_bifido_stacked_bar_plot.pdf', 
       plot = bifido_stacked_bar_plot, 
       width = 6, height = 4)


```
