---
title: "RAMP Correlation Network"
output: html_document
date: '2022-08-27'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Hmisc)
library(corrplot)
library(igraph)
library(viridis)
library(RColorBrewer)

mutate <- dplyr::mutate
```

```{r load_data_tables}
olink_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/OLINK/olink_results_scaled.csv') %>%
  rename(participant_timepoint = 1) %>%
  gather(key='cyto', value='npx', -participant_timepoint) %>%
  mutate(cyto = paste0('olink_', cyto)) %>%
  spread(key='cyto', value='npx') %>%
  filter(!grepl('4086', participant_timepoint))
  
metabolomics_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/METABOLOMICS/metabolomics_results_scaled.csv') %>%
  rename(participant_timepoint = 1)

phyloseq_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/16S/phyloseq_16S_results_scaled.csv') %>%
  rename(participant_timepoint = 1)
  
scfa_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/SCFA/scfa_results_scaled.csv') %>%
  rename(participant_timepoint = 1)


clinical_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/CLINICAL_DATA/clinical_results_scaled.csv') %>%
  rename(participant_timepoint = 1) %>%
  select(-c(clinical_total_cholesterol))

phosphoflow_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/PHOSPHOFLOW/phosphoflow_results_scaled.csv') %>%
  rename(participant_timepoint = 1)

dbcan_metagenomics_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/METAGENOMICS/dbcan_results_scaled.csv') %>%
  rename(participant_timepoint = 1)
```

```{r combine_data}
joined_data = left_join(metabolomics_data, olink_data) %>%
  # left_join(scfa_data) %>%
  left_join(clinical_data) %>%
  left_join(phosphoflow_data) %>%
  # left_join(dbcan_metagenomics_data) %>%
  left_join(phyloseq_data) %>%
  filter(!grepl('4011', participant_timepoint)) %>%
  gather(key='feature', value='value', -participant_timepoint) %>%
  rowwise() %>%
  mutate(participant_id = strsplit(participant_timepoint, '_')[[1]][1],
         timepoint = strsplit(participant_timepoint, '_')[[1]][2]) %>%
  select(-participant_timepoint) %>%
  spread(key='timepoint', value='value') %>%
  mutate(base_end = ifelse(grepl('microbiome', feature), (`4`-`2`), (`4`-`2`))) %>%
  select(participant_id, feature, base_end) %>%
  spread(key='feature', value='base_end') %>%
  column_to_rownames(var='participant_id')


color_df = colnames(joined_data) %>%
  as.data.frame() %>%
  rename(vertex = 1) %>%
  mutate(color = case_when(
    grepl('m_c18', vertex) ~ '#984EA3',  #377EB8
    grepl('microbiome', vertex) ~ '#4DAF4A',
    grepl('scfa', vertex) ~ '#FF7F00',
    grepl('clinical', vertex) ~ '#4DAF4A', 
    grepl('olink', vertex) ~ '#FF7F00',  #E41A1C
    grepl('phosphoflow', vertex) ~ '#A65628',
    grepl('EDTA', vertex) ~ '#ffffff'
  ))
```

```{r corr_network}

corr_matrix = rcorr(as.matrix(joined_data), type='pearson')

corr_r = corr_matrix$r
corr_r[lower.tri(corr_r, diag=T)] <- NA 
corr_r_filtered = corr_r

corr_p = corr_matrix$P
corr_p[lower.tri(corr_p, diag=T)] <- NA 

adjusted_p_mat = corr_p %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='P', -base) %>%
  mutate(p_adj = p.adjust(P)) %>%
  select(-P) %>%
  spread(key='compare', value='p_adj') %>%
  column_to_rownames(var='base') %>%
  as.matrix()

corr_r_filtered[adjusted_p_mat > 0.1] <- 0

# Make an Igraph object from this matrix:
network <- graph_from_adjacency_matrix(corr_r_filtered, weighted=T, mode="undirected", diag=F)
network <- delete.edges(network, E(network)[ weight < 0.0 ])
E(network)$weight[E(network)$weight <= 0] <- 0.0001

V(network)$color <- color_df$color

Isolated = which(igraph::degree(network)==0)
G2 = delete.vertices(network, Isolated)
```

```{r plot_igraph}
# Create igraph plot. Need to run chunk all at once for the saving to work. 
pdf("/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/metaanalysis_correlation_network_no_labels.pdf")
# Basic chart
igraph_plot = plot(G2,
     layout=layout.fruchterman.reingold,
     vertex.size=4,
     vertex.frame.width=0.4,
     # vertex.label=NA,
     vertex.label.cex=0.3,
     edge.color=ifelse(corr_matrix$r > 0, brewer.pal(7, 'Set3')[5],brewer.pal(7, 'Set3')[4]))

while (!is.null(dev.list()))  dev.off()
```

```{r }
metabolite_map = read_csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/metadata/metabolite_map.csv')

significant_cross_compares = corr_p %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='P', -base) %>%
  filter(!is.na(P)) %>%
  mutate(p_adj = p.adjust(P)) %>%
  filter(p_adj < 0.1) %>%
  rowwise() %>%
  mutate(base_pre = strsplit(base, '_')[[1]][1],
         compare_pre = strsplit(compare, '_')[[1]][1]) %>%
  filter(base_pre != compare_pre) %>%
  arrange(p_adj) %>%
  rowwise() %>%
  left_join(metabolite_map, by=c('base'='metabolite_name')) %>%
  select(base, compare, p_adj, inchikey)

p_value_compares = corr_p %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='P', -base) %>%
  filter(!is.na(P)) %>%
  rowwise() %>%
  mutate(base_pre = strsplit(base, '_')[[1]][1],
         compare_pre = strsplit(compare, '_')[[1]][1]) %>%
  filter(base_pre != compare_pre) %>%
  arrange(P) %>%
  rowwise() %>%
  left_join(metabolite_map, by=c('base'='metabolite_name')) %>%
  select(base, compare, P, inchikey)

all_cross_correlations = corr_r %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='R', -base) %>%
  arrange(desc(R)) %>%
  rowwise() %>%
  mutate(base_pre = strsplit(base, '_')[[1]][1],
         compare_pre = strsplit(compare, '_')[[1]][1]) %>%
  filter(base_pre != compare_pre) %>%
  rowwise() %>%
  left_join(metabolite_map, by=c('base'='metabolite_name')) %>%
  select(base, compare, R, inchikey)

significant_cross_compare_with_corr = significant_cross_compares %>%
  left_join(all_cross_correlations) %>%
  mutate(cyto_name = toupper(str_replace(compare, 'olink_', ''))) %>%
  filter(!grepl('clinical', base)) %>%
  mutate(inchikey = str_to_title(inchikey)) 

sig_compares = significant_cross_compares %>% 
  select(base, compare) %>%
  gather(key='type', value='name') %>%
  pull(name)

p_value_compares %>%
  filter(compare == 'microbiome_OTU.171' & base %in% sig_compares)

p_value_compares %>%
  filter(base == 'microbiome_OTU.171' & compare %in% sig_compares)
# %>%
#   mutate(inchikey = factor(inchikey, levels=c('Taurine', 'Ornithine', 'Glutamic Acid', 'Pyroglutamic Acid', 'Asparagine', 'Arginine', 'Inosine', 'Hypoxanthine', 'Guanosine', "5'-Methylthioadenosine", "Malic Acid")))

metabolite_by_cytokine_significance = ggplot(significant_cross_compare_with_corr, aes(x=cyto_name, y=inchikey, color=R)) +
  geom_point(aes(size=-log10(p_adj))) +
  scale_color_distiller(palette="RdBu", limits=c(-1, 1), direction=1) +
  scale_y_discrete(limits=rev) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -45, vjust = 1, hjust=0))

# ggsave('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/metaanalysis_mbx_olink_correlation_plot.pdf',
#        plot = metabolite_by_cytokine_significance, 
#        width = 6, height = 3
#        )


significant_cross_compare_with_corr

ggplot(joined_data %>% rownames_to_column(var='participant_id') %>% left_join(ramp_metadata_df), aes(x=microbiome_OTU.171, y=clinical_blood_glucose)) +
  geom_point() +
  stat_smooth(method='lm')
```


```{r shuo_will_data}
shuo_will_meta = read.csv('/home/mmcarter/user_data/scripts/220830_sh_wvt_nature_data/han_wvt_microbe_metadata.csv') %>%
  select(X, taxonomy, genus)
shuo_will_data = read.csv('/home/mmcarter/user_data/scripts/220830_sh_wvt_nature_data/han_wvt_microbe_foldchange.csv') %>%
  gather(key='metabolite', value='foldchange', -X) %>%
  left_join(shuo_will_meta) %>%
  filter(metabolite %in% (significant_cross_compare_with_corr %>% pull(base) %>% unique())) %>%
  filter(!is.na(foldchange)) %>% 
  filter(genus == 'Bifidobacterium') %>%
  arrange(desc(foldchange))


joined_data_all = left_join(metabolomics_data, olink_data) %>%
  left_join(phyloseq_data) %>%
  left_join(scfa_data) %>%
  left_join(clinical_data) %>%
  left_join(phosphoflow_data) %>%
  left_join(dbcan_metagenomics_data) %>%
  filter(!grepl('4011', participant_timepoint)) %>%
  gather(key='feature', value='value', -participant_timepoint) %>%
  rowwise() %>%
  mutate(participant_id = strsplit(participant_timepoint, '_')[[1]][1],
         timepoint = strsplit(participant_timepoint, '_')[[1]][2]) %>%
  select(-participant_timepoint) %>%
  spread(key='timepoint', value='value') %>%
  mutate(base_end = (`3`-`2`)) %>%
  select(participant_id, feature, base_end) %>%
  spread(key='feature', value='base_end') %>%
  column_to_rownames(var='participant_id')

ggplot(joined_data_all, aes(x=microbiome_OTU.191, y=m_c18p_0038)) +
  geom_point() +
  stat_smooth(method='lm')

summary(lm(m_c18p_0038 ~ microbiome_OTU.191, data=joined_data_all))

corr_matrix = rcorr(as.matrix(joined_data_all), type='pearson')

corr_r = corr_matrix$r
corr_r[lower.tri(corr_r, diag=T)] <- NA 

corr_p = corr_matrix$P
corr_p[lower.tri(corr_p, diag=T)] <- NA 

adjusted_p_mat = corr_matrix$P %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='P', -base) 

sig_hits = c('microbiome_OTU.191', 
             'm_c18p_0038',
             'm_c18p_0098',
             'm_c18p_0049',
             'm_c18p_0032',
             'm_c18p_0067',
             'm_c18p_0447',
             'st1a1',
             'tnfsf14',
             'sirt2',
             'osm',
             'mcp3')

adjusted_p_mat %>%
  filter(base %in% sig_hits | compare %in% sig_hits) %>%
  filter(!is.na(P)) %>%
  mutate(p_adj = p.adjust(P)) %>%
  arrange(p_adj) %>%
  filter(p_adj < 0.1)
 

# %>%
#   select(-P) %>%
#   spread(key='compare', value='p_adj') %>%
#   column_to_rownames(var='base') %>%
#   as.matrix()

ggplot(joined_data_all %>% rownames_to_column(var='participant_id') %>% left_join(ramp_metadata_df), aes(x=m_c18p_0098, y=m_c18p_0447)) +
  geom_point() +
  geom_text(aes(label=participant_id)) +
  stat_smooth(method='lm')
```


