---
title: "RAMP Correlation Network"
output: html_document
date: '2022-08-27'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Hmisc)
library(corrplot)
library(igraph)
library(viridis)
library(RColorBrewer)

mutate <- dplyr::mutate
```

```{r load_data_tables}
olink_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/OLINK/olink_results_scaled.csv') %>%
  rename(participant_timepoint = 1) %>%
  gather(key='cyto', value='npx', -participant_timepoint) %>%
  mutate(cyto = paste0('olink_', cyto)) %>%
  spread(key='cyto', value='npx') %>%
  filter(!grepl('4086', participant_timepoint))
  
metabolomics_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/METABOLOMICS/metabolomics_results_scaled.csv') %>%
  rename(participant_timepoint = 1)

phyloseq_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/16S/phyloseq_16S_results_scaled.csv') %>%
  rename(participant_timepoint = 1)
  
scfa_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/SCFA/scfa_results_scaled.csv') %>%
  rename(participant_timepoint = 1)


clinical_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/CLINICAL_DATA/clinical_results_scaled.csv') %>%
  rename(participant_timepoint = 1) %>%
  select(-c(clinical_total_cholesterol))

phosphoflow_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/PHOSPHOFLOW/phosphoflow_results_scaled.csv') %>%
  rename(participant_timepoint = 1)

dbcan_metagenomics_data = read.csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/data/METAGENOMICS/dbcan_results_scaled.csv') %>%
  rename(participant_timepoint = 1)
```

```{r combine_data}
joined_data = left_join(metabolomics_data, olink_data) %>%
  left_join(phyloseq_data) %>%
  left_join(clinical_data) %>%
  filter(!grepl('4011', participant_timepoint)) %>%
  gather(key='feature', value='value', -participant_timepoint) %>%
  rowwise() %>%
  mutate(participant_id = strsplit(participant_timepoint, '_')[[1]][1],
         timepoint = strsplit(participant_timepoint, '_')[[1]][2]) %>%
  select(-participant_timepoint) %>%
  spread(key='timepoint', value='value') %>%
  mutate(base_end = ifelse(grepl('microbiome', feature), (`4`-`2`), (`4`-`2`))) %>%
  select(participant_id, feature, base_end) %>%
  spread(key='feature', value='base_end') %>%
  column_to_rownames(var='participant_id')


color_df = colnames(joined_data) %>%
  as.data.frame() %>%
  rename(vertex = 1) %>%
  mutate(color = case_when(
    grepl('m_c18', vertex) ~ '#984EA3',  #377EB8
    grepl('clinical', vertex) ~ '#4DAF4A', 
    grepl('microbiome', vertex) ~ '#444444',
    grepl('olink', vertex) ~ '#FF7F00',  #E41A1C
  ))
color_df_rownames = color_df %>% 
  column_to_rownames(var='vertex')
```

```{r corr_network}

corr_matrix = rcorr(as.matrix(joined_data), type='pearson')

corr_r = corr_matrix$r
corr_r[lower.tri(corr_r, diag=T)] <- NA 
corr_r_filtered = corr_r

corr_p = corr_matrix$P
corr_p[lower.tri(corr_p, diag=T)] <- NA 

adjusted_p_mat = corr_p %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='P', -base) %>%
  mutate(p_adj = p.adjust(P)) %>%
  select(-P) %>%
  spread(key='compare', value='p_adj') %>%
  column_to_rownames(var='base') %>%
  as.matrix()

# corr_r_filtered[adjusted_p_mat > 0.1] <- 0
corr_r_filtered[adjusted_p_mat > 0.05] <- 0

corr_r_filtered_long = corr_r_filtered %>% 
  as.data.frame() %>% 
  rownames_to_column(var='feature1') %>% 
  gather(key='feature2', value='R', -feature1) %>% 
  filter(!is.na(R)) %>% 
  filter(abs(R) > 0) 

corr_r_filtered_long

t.graph=graph.data.frame(corr_r_filtered_long,directed=F)
E(t.graph)$color <- ifelse(E(t.graph)$R > 0, brewer.pal(7, 'Set3')[5],brewer.pal(7, 'Set3')[4])


color_vect = names(V(t.graph) %>% as.list()) %>% 
  as.data.frame() %>% 
  rename(vertex = 1) %>%
  mutate(color = case_when(
    grepl('m_c18', vertex) ~ '#3D82C4',  #377EB8
    grepl('clinical', vertex) ~ '#014D41', 
    grepl('olink', vertex) ~ '#D81B61'
  )) %>% 
  pull(color)

V(t.graph)$color = color_vect

Isolated = which(igraph::degree(t.graph)==0)
G2 = delete.vertices(t.graph, Isolated)

# # Make an Igraph object from this matrix:
# network <- graph_from_adjacency_matrix(corr_r_filtered, weighted=T, mode="undirected", diag=F)
# network <- delete.edges(network, E(network)[ weight < 0.0 ])
# E(network)$weight[E(network)$weight <= 0] <- 0.0001
# 
# gg <- nexus.get("Davis")
# 
# V(network)$color <- color_df$color
# network$
# a = E(network)
# a$V3
# Isolated = which(igraph::degree(network)==0)
# G2 = delete.vertices(network, Isolated)
```

```{r plot_igraph}
# Create igraph plot. Need to run chunk all at once for the saving to work. 
# pdf("/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/metaanalysis_correlation_network_no_labels.pdf")
pdf("/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/metaanalysis_correlation_network_no_labels_v3.pdf")
# Basic chart
igraph_plot = plot(G2,
     layout=layout.fruchterman.reingold,
     vertex.size=4,
     vertex.frame.width=0.4,
     # vertex.label=NA,
     vertex.label.cex=0.3,
     edge.color=ifelse(corr_matrix$r > 0, brewer.pal(7, 'Set3')[5],brewer.pal(7, 'Set3')[4]))

set.seed(126)

igraph_plot2 = plot(G2,
     layout=layout.fruchterman.reingold,
     vertex.size=4,
     vertex.frame.width=0.4,
     vertex.label=NA)
     # vertex.label.cex=0.3)

while (!is.null(dev.list()))  dev.off()


```

```{r }
metabolite_map = read_csv('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/metadata/metabolite_map.csv')

significant_cross_compares = corr_p %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='P', -base) %>%
  filter(!is.na(P)) %>%
  mutate(p_adj = p.adjust(P)) %>%
  filter(p_adj < 0.1) %>%
  rowwise() %>%
  mutate(base_pre = strsplit(base, '_')[[1]][1],
         compare_pre = strsplit(compare, '_')[[1]][1]) %>%
  filter(base_pre != compare_pre) %>%
  arrange(p_adj) %>%
  rowwise() %>%
  left_join(metabolite_map, by=c('base'='metabolite_name')) %>%
  select(base, compare, p_adj, inchikey)

p_value_compares = corr_p %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='P', -base) %>%
  filter(!is.na(P)) %>%
  rowwise() %>%
  mutate(base_pre = strsplit(base, '_')[[1]][1],
         compare_pre = strsplit(compare, '_')[[1]][1]) %>%
  filter(base_pre != compare_pre) %>%
  arrange(P) %>%
  rowwise() %>%
  left_join(metabolite_map, by=c('base'='metabolite_name')) %>%
  select(base, compare, P, inchikey)

all_cross_correlations = corr_r %>%
  as.data.frame() %>%
  rownames_to_column(var='base') %>%
  gather(key='compare', value='R', -base) %>%
  arrange(desc(R)) %>%
  rowwise() %>%
  mutate(base_pre = strsplit(base, '_')[[1]][1],
         compare_pre = strsplit(compare, '_')[[1]][1]) %>%
  filter(base_pre != compare_pre) %>%
  rowwise() %>%
  left_join(metabolite_map, by=c('base'='metabolite_name')) %>%
  select(base, compare, R, inchikey)

significant_cross_compare_with_corr = significant_cross_compares %>%
  left_join(all_cross_correlations) %>%
  mutate(cyto_name = toupper(str_replace(compare, 'olink_', ''))) %>%
  filter(!grepl('clinical', base)) %>%
  mutate(inchikey = str_to_title(inchikey)) 

sig_compares = significant_cross_compares %>% 
  select(base, compare) %>%
  gather(key='type', value='name') %>%
  pull(name)

metabolite_by_cytokine_significance = ggplot(significant_cross_compare_with_corr, aes(x=cyto_name, y=inchikey, color=R)) +
  geom_point(aes(size=-log10(p_adj))) +
  scale_color_distiller(palette="RdBu", limits=c(-1, 1), direction=1) +
  scale_y_discrete(limits=rev) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -45, vjust = 1, hjust=0))

ggsave('/home/mmcarter/user_data/Projects/RAMP/RAMP_study_analysis/plots/metaanalysis_mbx_olink_correlation_plot.pdf',
       plot = metabolite_by_cytokine_significance,
       width = 6, height = 3
       )
```


